File: python/src/qti_neon/relay.py
"""Neon relay server."""

from __future__ import annotations

import enum
import logging
import time
import threading
from collections import deque
from dataclasses import dataclass

from ._config import NeonConfig
from ._protocol import (
    NeonPacket,
    PacketType,
    PacketHeader,
    ConnectAccept,
    ConnectDeny,
    DisconnectNotice,
)
from ._rate_limiter import _RateLimiter
from ._session import _SessionManager
from ._socket import _NeonSocket

logger = logging.getLogger(__name__)


@dataclass
class _PendingConnection:
    session_id: int
    name: str
    requested_at: float


@dataclass
class _PendingReconnect:
    session_id: int
    client_id: int
    new_address: tuple[str, int]
    requested_at: float


class _State(enum.Enum):
    CREATED = "CREATED"
    RUNNING = "RUNNING"
    STOPPED = "STOPPED"
    FAILED = "FAILED"


class NeonRelay:
    """UDP relay server for the Neon multiplayer protocol.

    Routes packets between session hosts and clients. The relay is stateless
    with respect to game logic — it only understands protocol lifecycle packets
    and routes everything else by destination ID.

    Typical usage:

    ```python
    import threading
    from qti_neon import NeonRelay, NeonConfig

    relay = NeonRelay("0.0.0.0", NeonConfig())
    thread = threading.Thread(target=relay.start_and_run, daemon=True)
    thread.start()

    # … later …
    relay.stop()
    ```
    """

    def __init__(self, bind_address: str, config: NeonConfig | None = None) -> None:
        """Create a relay bound to *bind_address* on ``config.relay_port``.

        Args:
            bind_address: IP address to bind (e.g. ``"0.0.0.0"`` for all interfaces).
            config: Protocol configuration. Uses defaults when ``None``.

        Raises:
            OSError: if the socket cannot be bound.
        """
        self._config = config or NeonConfig()
        self._socket = _NeonSocket(
            bind_address=(bind_address, self._config.relay_port),
            config=self._config,
        )
        self._session_manager = _SessionManager()
        self._rate_limiters: dict[tuple[str, int], _RateLimiter] = {}
        # pending_connections and pending_by_session are only touched from the
        # relay processing thread — no locking required.
        self._pending_connections: dict[tuple[str, int], _PendingConnection] = {}
        self._pending_by_session: dict[int, deque[tuple[str, int]]] = {}
        self._pending_reconnects: dict[str, _PendingReconnect] = {}
        self._last_cleanup: float = time.monotonic()
        self._state = _State.CREATED
        self._state_lock = threading.Lock()

    # ------------------------------------------------------------------ #
    # Lifecycle                                                            #
    # ------------------------------------------------------------------ #

    def start(self) -> None:
        """Initialise the relay without entering the processing loop.

        Call `start_and_run` instead for typical use.

        Raises:
            RuntimeError: if called from a state other than ``CREATED``.
        """
        with self._state_lock:
            if self._state is not _State.CREATED:
                raise RuntimeError(f"Cannot start from state: {self._state.value}")
            self._state = _State.RUNNING
        if self._config.is_dtls_enabled:
            self._socket.enable_server_dtls(self._config.dtls_config)  # type: ignore[arg-type]
        logger.info("NeonRelay started on port %d", self._config.relay_port)

    def stop(self) -> None:
        """Signal the relay to stop and close the socket.

        Raises:
            RuntimeError: if the relay is not running.
        """
        with self._state_lock:
            if self._state is not _State.RUNNING:
                raise RuntimeError(f"Cannot stop from state: {self._state.value}")
            self._state = _State.STOPPED
        self._socket.close()

    def start_and_run(self) -> None:
        """Start the relay and block in the processing loop until `stop` is called.

        Intended to run in a dedicated thread:

        ```python
        threading.Thread(target=relay.start_and_run, daemon=True).start()
        ```

        Raises:
            OSError: if socket initialisation fails.
        """
        self.start()
        self._run()

    def _run(self) -> None:
        while self._state is _State.RUNNING:
            try:
                self._process_packets()
                self._perform_cleanup()
                time.sleep(self._config.relay_main_loop_sleep_ms / 1000.0)
            except InterruptedError:
                break
            except OSError as exc:
                if not self._socket.is_closed:
                    logger.warning("IO error in relay loop: %s", exc)
                break

    @property
    def is_running(self) -> bool:
        """``True`` while the relay processing loop is active."""
        return self._state is _State.RUNNING

    def local_address(self) -> tuple[str, int]:
        """Return the local ``(host, port)`` the relay is bound to."""
        return self._socket.local_address()

    def __enter__(self) -> "NeonRelay":
        return self

    def __exit__(self, *_: object) -> None:
        if self._state is _State.RUNNING:
            self.stop()

    # ------------------------------------------------------------------ #
    # Packet processing                                                    #
    # ------------------------------------------------------------------ #

    def _process_packets(self) -> None:
        while True:
            result = self._socket.receive_packet()
            if result is None:
                break
            packet, source = result
            self._session_manager.update_last_seen(source)

            limiter = self._rate_limiters.get(source)
            if limiter is None:
                limiter = _RateLimiter(self._config.max_packets_per_second)
                self._rate_limiters[source] = limiter
            if not limiter.allow_packet():
                logger.debug("Rate limited: %s", source)
                continue

            self._handle_packet(packet, source)

    def _handle_packet(self, packet: NeonPacket, source: tuple[str, int]) -> None:
        payload = packet.payload
        match payload:
            case _HostRegister():
                self._handle_host_register(payload, source)
            case _ConnectRequest():
                self._handle_connect_request(payload, source, packet.header)
            case ConnectAccept():
                self._handle_connect_accept(payload, source, packet.header)
            case _ReconnectRequest():
                self._handle_reconnect_request(payload, source)
            case DisconnectNotice():
                self._handle_disconnect_notice(source, packet.header)
            case _:
                self._route_packet(packet, source)

    def _handle_host_register(
        self, req: "_HostRegister", source: tuple[str, int]
    ) -> None:
        if req.session_id <= 0:
            logger.warning("Rejected HOST_REGISTER: invalid session_id=%d", req.session_id)
            return
        self._session_manager.register_host(req.session_id, source)
        logger.info("Host registered for session %d from %s", req.session_id, source)
        accept = ConnectAccept(
            client_id=1,
            session_id=req.session_id,
            token=req.host_token,
        )
        self._socket.send_packet(
            NeonPacket.create(PacketType.CONNECT_ACCEPT, 0, 0, 1, accept),
            source,
        )

    def _handle_connect_request(
        self,
        req: "_ConnectRequest",
        source: tuple[str, int],
        header: PacketHeader,
    ) -> None:
        host = self._session_manager.get_host(req.session_id)
        if host is None:
            self._send_deny(source, "Session not found")
            return
        if self._session_manager.get_client_count(req.session_id) >= self._config.max_clients_per_session:
            self._send_deny(source, "Session full")
            return
        if len(self._pending_connections) >= self._config.max_pending_connections:
            self._send_deny(source, "Relay busy")
            return

        self._pending_connections[source] = _PendingConnection(
            session_id=req.session_id,
            name=req.name,
            requested_at=time.monotonic(),
        )
        queue = self._pending_by_session.setdefault(req.session_id, deque())
        queue.append(source)

        self._socket.send_packet(
            NeonPacket.create(
                PacketType.CONNECT_REQUEST,
                header.sequence,
                header.client_id,
                1,
                req,
            ),
            host,
        )

    def _handle_connect_accept(
        self,
        accept: ConnectAccept,
        source: tuple[str, int],
        header: PacketHeader,
    ) -> None:
        client_id = accept.client_id
        session_id = accept.session_id

        if client_id == 1:
            # Relay acknowledges its own host registration — ignore.
            logger.debug("Ignoring CONNECT_ACCEPT(client_id=1) from %s", source)
            return

        reconnect_key = f"{session_id}:{client_id}"
        reconnect = self._pending_reconnects.pop(reconnect_key, None)
        if reconnect is not None:
            self._session_manager.update_peer_address(
                session_id, client_id, reconnect.new_address
            )
            self._socket.send_packet(
                NeonPacket.create(
                    PacketType.CONNECT_ACCEPT,
                    header.sequence,
                    header.client_id,
                    client_id,
                    accept,
                ),
                reconnect.new_address,
            )
            logger.info(
                "Reconnect confirmed: client %d session %d", client_id, session_id
            )
            return

        queue = self._pending_by_session.get(session_id)
        client_addr = queue.popleft() if queue else None
        if client_addr is None:
            logger.warning(
                "CONNECT_ACCEPT for session %d but no pending clients", session_id
            )
            return
        if not queue:
            self._pending_by_session.pop(session_id, None)

        self._pending_connections.pop(client_addr, None)
        self._session_manager.register_peer(session_id, client_id, client_addr)
        self._socket.send_packet(
            NeonPacket.create(
                PacketType.CONNECT_ACCEPT,
                header.sequence,
                header.client_id,
                client_id,
                accept,
            ),
            client_addr,
        )
        logger.info("Client %d joined session %d", client_id, session_id)

    def _handle_reconnect_request(
        self, req: "_ReconnectRequest", source: tuple[str, int]
    ) -> None:
        host = self._session_manager.get_host(req.session_id)
        if host is None:
            self._send_deny(source, "Session not found")
            return
        key = f"{req.session_id}:{req.previous_client_id}"
        self._pending_reconnects[key] = _PendingReconnect(
            session_id=req.session_id,
            client_id=req.previous_client_id,
            new_address=source,
            requested_at=time.monotonic(),
        )
        self._socket.send_packet(
            NeonPacket.create(
                PacketType.RECONNECT_REQUEST,
                0,
                req.previous_client_id,
                1,
                req,
            ),
            host,
        )

    def _handle_disconnect_notice(
        self, source: tuple[str, int], header: PacketHeader
    ) -> None:
        session_id = self._session_manager.get_session_for_peer(source)
        if session_id is not None:
            notice_bytes = NeonPacket.create(
                PacketType.DISCONNECT_NOTICE,
                header.sequence,
                header.client_id,
                0,
                DisconnectNotice(),
            ).to_bytes()
            for peer in self._session_manager.get_all_peers_except(session_id, source):
                self._socket.send(notice_bytes, peer)

        self._session_manager.remove_peer(source)
        self._pending_connections.pop(source, None)
        self._rate_limiters.pop(source, None)
        self._socket.remove_dtls_session(source)

    def _route_packet(self, packet: NeonPacket, source: tuple[str, int]) -> None:
        session_id = self._session_manager.get_session_for_peer(source)
        if session_id is None:
            logger.debug("Unroutable from %s: not in any session", source)
            return

        dest_id = packet.header.destination_id
        raw = packet.to_bytes()

        if dest_id == 0:
            peers = self._session_manager.get_all_peers_except(session_id, source)
            if not peers:
                logger.debug("Unroutable from %s: no peers to broadcast to", source)
                return
            for peer in peers:
                self._socket.send(raw, peer)
        else:
            dest = self._session_manager.get_peer_address(session_id, dest_id)
            if dest is None:
                logger.debug(
                    "Unroutable from %s: destination client %d not found", source, dest_id
                )
                return
            self._socket.send(raw, dest)

    def _send_deny(self, dest: tuple[str, int], reason: str) -> None:
        self._socket.send_packet(
            NeonPacket.create(PacketType.CONNECT_DENY, 0, 0, 0, ConnectDeny(reason)),
            dest,
        )

    # ------------------------------------------------------------------ #
    # Cleanup                                                              #
    # ------------------------------------------------------------------ #

    def _perform_cleanup(self) -> None:
        now = time.monotonic()
        if now - self._last_cleanup < self._config.relay_cleanup_interval_ms / 1000.0:
            return
        self._last_cleanup = now

        self._session_manager.remove_stale(self._config.relay_client_timeout_ms)

        cutoff = now - self._config.relay_client_timeout_ms / 1000.0
        stale_conns = [
            addr
            for addr, conn in self._pending_connections.items()
            if conn.requested_at < cutoff
        ]
        for addr in stale_conns:
            conn = self._pending_connections.pop(addr)
            queue = self._pending_by_session.get(conn.session_id)
            if queue is not None:
                try:
                    queue.remove(addr)
                except ValueError:
                    pass
                if not queue:
                    self._pending_by_session.pop(conn.session_id, None)

        stale_rc = [
            k for k, rc in self._pending_reconnects.items() if rc.requested_at < cutoff
        ]
        for k in stale_rc:
            del self._pending_reconnects[k]

        if len(self._rate_limiters) > self._config.max_rate_limiters:
            self._rate_limiters.clear()


# Local imports to avoid forward-reference issues in match statements
from ._protocol import HostRegister as _HostRegister
from ._protocol import ConnectRequest as _ConnectRequest
from ._protocol import ReconnectRequest as _ReconnectRequest

File: js-ts/src/relay.ts
import { NeonConfig } from "./_config";
import { NeonSocket, Address, addressKey } from "./_socket";
import { SessionManager } from "./_session";
import { RateLimiter } from "./_rate_limiter";
import {
  NeonPacket,
  PacketType,
  makePacket,
  serializePacket,
  isConnectRequest,
  isConnectAccept,
  isConnectDeny,
  isHostRegister,
  isReconnectRequest,
  isDisconnectNotice,
  ConnectAcceptPayload,
  ConnectDenyPayload,
  DisconnectNoticePayload,
} from "./_protocol";

interface PendingConnection {
  sessionId: number;
  requestedAt: number;
}

interface PendingReconnect {
  sessionId: number;
  clientId: number;
  newAddress: Address;
  requestedAt: number;
}

enum State {
  CREATED = "CREATED",
  RUNNING = "RUNNING",
  STOPPED = "STOPPED",
}

export class NeonRelay {
  private config: NeonConfig;
  private socket: NeonSocket;
  private sessionManager = new SessionManager();
  private rateLimiters = new Map<string, RateLimiter>();
  private pendingConnections = new Map<string, PendingConnection>();
  private pendingBySession = new Map<number, Address[]>();
  private pendingReconnects = new Map<string, PendingReconnect>();
  private lastCleanup = Date.now();
  private cleanupTimer: NodeJS.Timeout | null = null;
  private state = State.CREATED;

  constructor(bindAddress: string, config?: NeonConfig) {
    this.config = config ?? new NeonConfig();
    const [host, portStr] = bindAddress.includes(":")
      ? [bindAddress.slice(0, bindAddress.lastIndexOf(":")), bindAddress.slice(bindAddress.lastIndexOf(":") + 1)]
      : [bindAddress, String(this.config.relayPort)];
    this.socket = new NeonSocket({ address: host, port: parseInt(portStr, 10) }, this.config);
    this.socket.on("packet", (pkt: NeonPacket, addr: Address) => this._onPacket(pkt, addr));
  }

  async start(): Promise<void> {
    if (this.state !== State.CREATED) throw new Error(`Cannot start from state: ${this.state}`);
    await this.socket.waitBound();
    if (this.config.isDtlsEnabled) {
      this.socket.enableServerDtls(this.config.dtlsConfig!);
    }
    this.state = State.RUNNING;
    this.cleanupTimer = setInterval(() => this._performCleanup(), this.config.relayCleanupIntervalMs);
    if (this.cleanupTimer.unref) this.cleanupTimer.unref();
  }

  async startAndRun(): Promise<void> {
    await this.start();
  }

  stop(): void {
    if (this.state !== State.RUNNING) throw new Error(`Cannot stop from state: ${this.state}`);
    this.state = State.STOPPED;
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
    this.socket.close();
  }

  get isRunning(): boolean {
    return this.state === State.RUNNING;
  }

  localAddress(): Address {
    return this.socket.localAddress();
  }

  private _onPacket(pkt: NeonPacket, source: Address): void {
    if (this.state !== State.RUNNING) return;

    this.sessionManager.updateLastSeen(source);

    const key = addressKey(source);
    let limiter = this.rateLimiters.get(key);
    if (!limiter) {
      limiter = new RateLimiter(this.config.maxPacketsPerSecond);
      this.rateLimiters.set(key, limiter);
    }
    if (!limiter.allowPacket()) return;

    this._handlePacket(pkt, source);
  }

  private _handlePacket(pkt: NeonPacket, source: Address): void {
    const p = pkt.payload;
    if (isHostRegister(p)) {
      this._handleHostRegister(p.sessionId, p.hostToken, source);
    } else if (isConnectRequest(p)) {
      this._handleConnectRequest(p.sessionId, p, pkt.header, source);
    } else if (isConnectAccept(p)) {
      this._handleConnectAccept(p, pkt.header, source);
    } else if (isConnectDeny(p)) {
      this._handleConnectDeny(p, pkt.header, source);
    } else if (isReconnectRequest(p)) {
      this._handleReconnectRequest(p.sessionId, p.previousClientId, p, source);
    } else if (isDisconnectNotice(p)) {
      this._handleDisconnectNotice(source, pkt.header);
    } else {
      this._routePacket(pkt, source);
    }
  }

  private _handleHostRegister(sessionId: number, hostToken: bigint, source: Address): void {
    if (sessionId <= 0) return;
    this.sessionManager.registerHost(sessionId, source);
    const accept: ConnectAcceptPayload = { clientId: 1, sessionId, token: hostToken };
    this.socket.sendPacket(makePacket(PacketType.CONNECT_ACCEPT, 0, 0, 1, accept), source);
  }

  private _handleConnectRequest(
    sessionId: number,
    payload: NeonPacket["payload"],
    header: NeonPacket["header"],
    source: Address
  ): void {
    const host = this.sessionManager.getHost(sessionId);
    if (!host) {
      this._sendDeny(source, "Session not found");
      return;
    }
    if (this.sessionManager.getClientCount(sessionId) >= this.config.maxClientsPerSession) {
      this._sendDeny(source, "Session full");
      return;
    }
    if (this.pendingConnections.size >= this.config.maxPendingConnections) {
      this._sendDeny(source, "Relay busy");
      return;
    }

    const key = addressKey(source);
    this.pendingConnections.set(key, { sessionId, requestedAt: Date.now() });

    let queue = this.pendingBySession.get(sessionId);
    if (!queue) {
      queue = [];
      this.pendingBySession.set(sessionId, queue);
    }
    queue.push(source);

    this.socket.sendPacket(
      makePacket(PacketType.CONNECT_REQUEST, header.sequence, header.clientId, 1, payload),
      host
    );
  }

  private _handleConnectAccept(accept: ConnectAcceptPayload, header: NeonPacket["header"], source: Address): void {
    const { clientId, sessionId } = accept;

    if (clientId === 1) return;

    const reconnectKey = `${sessionId}:${clientId}`;
    const reconnect = this.pendingReconnects.get(reconnectKey);
    if (reconnect) {
      this.pendingReconnects.delete(reconnectKey);
      this.sessionManager.updatePeerAddress(sessionId, clientId, reconnect.newAddress);
      this.socket.sendPacket(
        makePacket(PacketType.CONNECT_ACCEPT, header.sequence, header.clientId, clientId, accept),
        reconnect.newAddress
      );
      return;
    }

    const queue = this.pendingBySession.get(sessionId);
    const clientAddr = queue?.shift();
    if (!clientAddr) return;
    if (queue && queue.length === 0) this.pendingBySession.delete(sessionId);

    this.pendingConnections.delete(addressKey(clientAddr));
    this.sessionManager.registerPeer(sessionId, clientId, clientAddr);
    this.socket.sendPacket(
      makePacket(PacketType.CONNECT_ACCEPT, header.sequence, header.clientId, clientId, accept),
      clientAddr
    );
  }

  private _handleConnectDeny(deny: ConnectDenyPayload, header: NeonPacket["header"], source: Address): void {
    const sessionId = this.sessionManager.getSessionForPeer(source);
    if (sessionId === undefined) return;

    const queue = this.pendingBySession.get(sessionId);
    const clientAddr = queue?.shift();
    if (!clientAddr) return;
    if (queue && queue.length === 0) this.pendingBySession.delete(sessionId);

    this.pendingConnections.delete(addressKey(clientAddr));
    this.socket.sendPacket(
      makePacket(PacketType.CONNECT_DENY, header.sequence, 1, 0, deny),
      clientAddr
    );
  }

  private _handleReconnectRequest(
    sessionId: number,
    previousClientId: number,
    payload: NeonPacket["payload"],
    source: Address
  ): void {
    const host = this.sessionManager.getHost(sessionId);
    if (!host) {
      this._sendDeny(source, "Session not found");
      return;
    }
    const key = `${sessionId}:${previousClientId}`;
    this.pendingReconnects.set(key, {
      sessionId,
      clientId: previousClientId,
      newAddress: source,
      requestedAt: Date.now(),
    });
    this.socket.sendPacket(makePacket(PacketType.RECONNECT_REQUEST, 0, previousClientId, 1, payload), host);
  }

  private _handleDisconnectNotice(source: Address, header: NeonPacket["header"]): void {
    const sessionId = this.sessionManager.getSessionForPeer(source);
    if (sessionId !== undefined) {
      const notice: DisconnectNoticePayload = { _tag: "disconnect" };
      const raw = serializePacket(makePacket(PacketType.DISCONNECT_NOTICE, header.sequence, header.clientId, 0, notice));
      for (const peer of this.sessionManager.getAllPeersExcept(sessionId, source)) {
        this.socket.send(raw, peer);
      }
    }
    this.sessionManager.removePeer(source);
    this.pendingConnections.delete(addressKey(source));
    this.rateLimiters.delete(addressKey(source));
    this.socket.removeDtlsSession(source);
  }

  private _routePacket(pkt: NeonPacket, source: Address): void {
    const sessionId = this.sessionManager.getSessionForPeer(source);
    if (sessionId === undefined) return;

    const destId = pkt.header.destinationId;
    const raw = serializePacket(pkt);

    if (destId === 0) {
      for (const peer of this.sessionManager.getAllPeersExcept(sessionId, source)) {
        this.socket.send(raw, peer);
      }
    } else {
      const dest = this.sessionManager.getPeerAddress(sessionId, destId);
      if (dest) this.socket.send(raw, dest);
    }
  }

  private _sendDeny(dest: Address, reason: string): void {
    const p: ConnectDenyPayload = { reason };
    this.socket.sendPacket(makePacket(PacketType.CONNECT_DENY, 0, 0, 0, p), dest);
  }

  private _performCleanup(): void {
    if (this.state !== State.RUNNING) return;
    this.sessionManager.removeStale(this.config.relayClientTimeoutMs);

    const cutoff = Date.now() - this.config.relayClientTimeoutMs;
    for (const [key, conn] of this.pendingConnections) {
      if (conn.requestedAt < cutoff) {
        this.pendingConnections.delete(key);
        const queue = this.pendingBySession.get(conn.sessionId);
        if (queue) {
          const idx = queue.findIndex((a) => addressKey(a) === key);
          if (idx >= 0) queue.splice(idx, 1);
          if (queue.length === 0) this.pendingBySession.delete(conn.sessionId);
        }
      }
    }

    for (const [key, rc] of this.pendingReconnects) {
      if (rc.requestedAt < cutoff) this.pendingReconnects.delete(key);
    }

    if (this.rateLimiters.size > this.config.maxRateLimiters) {
      this.rateLimiters.clear();
    }
  }
}
